### PR TITLE
Make IOutputConsole async

### LINKED ISSUE #9268: Make IOutputConsole async
IOutputConsole is currently sync. The reason for that is that it was original designed when PMC was the primary Visual Studio scenario & async code was not as common.

Many of our codepaths currently invoking the output console are async and with https://github.com/nuget/home/issues/8673 that's becoming even more important. 

Refactor IOutputConsole to contain only async methods.

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/nuget/home/issues/9268
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 

IOutputConsole is currently sync. The reason for that is that it was original designed when PMC was the primary Visual Studio scenario & async code was not as common.

Many of our codepaths currently invoking the output console are async and with #8673 that's becoming even more important.

Refactor IOutputConsole to contain only sync methods.

In this change, I have removed a bunch of unnecessary UI thread switches.

## Testing/Validation

Tests Added: No  
Reason for not adding tests: Current tests cover these scenarios. There's no experience changes beyond this.
Validation:  Manual + automation

### CODE REVIEW

File: src/NuGet.Clients/NuGet.Console/Console/ConsoleDispatcher.cs [hunk 1]
```diff
@@ -323,7 +317,7 @@ protected Dispatcher(ConsoleDispatcher parentDispatcher)
             [SuppressMessage("Microsoft.Globalization", "CA1303")]
             protected void PromptNewLine()
             {
-                WpfConsole.Write(WpfConsole.Host.Prompt + (char)32); // 32 is the space
+                NuGetUIThreadHelper.JoinableTaskFactory.Run(() => WpfConsole.WriteAsync(WpfConsole.Host.Prompt + (char)32)); // 32 is the space
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.Console/Console/ConsoleDispatcher.cs]:
```suggestion
                NuGetUIThreadHelper.JoinableTaskFactory.Run(() => WpfConsole.WriteAsync(WpfConsole.Host.Prompt + ' ');
```

**Review by zivkan — APPROVED** [review]
I believe E2E and/or Apex tests use console output to assert the test worked as intended.  Do we need to modify them to avoid flaky tests in case the test assertion runs before the console write async task finishes?

**Comment by nkolev92** [comment]
Yeah, I adjusted the E2E tests.

**Comment by nkolev92** [comment]
All blocking end to end scenarios remain blocking, so no need to adjust any of the tests.

File: src/NuGet.Clients/NuGet.Console/OutputConsole/OutputConsole.cs [hunk 2]
```diff
@@ -85,65 +79,30 @@ public void Write(string text)
...
-        public void WriteBackspace()
-        {
-            throw new NotSupportedException();
-        }
-
-        public void WriteLine(string text)
-        {
-            Write(text + Environment.NewLine);
-        }
+            _vsUiShell.FindToolWindow(0,
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.Console/OutputConsole/OutputConsole.cs:90]: Add argument names to this call

File: src/NuGet.Clients/NuGet.Console/OutputConsole/SharedOutputConsole.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    internal abstract class SharedOutputConsole : IOutputConsole
+    {
+        // The next 3 methods are the ones that need to be overriden
+        public abstract Task ActivateAsync();
+
+        public abstract Task ClearAsync();
+
+        public abstract Task WriteAsync(string text);
+
+        public int ConsoleWidth => 120;
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.Console/OutputConsole/SharedOutputConsole.cs]: Put the property at the beginning of the class

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs [hunk 4]
```diff
@@ -633,12 +633,12 @@ protected void ReportError(Exception exception)
...
         private void WriteErrorLine(string message)
         {
-            ActiveConsole?.Write(message + Environment.NewLine, Colors.White, Colors.DarkRed);
+            ActiveConsole?.WriteAsync(message + Environment.NewLine, Colors.White, Colors.DarkRed);
         }
 
         private void WriteLine(string message = "")
         {
-            ActiveConsole?.WriteLine(message);
+            ActiveConsole?.WriteLineAsync(message);
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: Do you need to wait to finish this async method? The same applies for the line modified above.

**Review by dominoFire — APPROVED** [review]
async/await issue in src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs and some coding guidelines nits. Overall, looks good.

File: src/NuGet.Clients/NuGet.Console/OutputConsole/OutputConsole.cs (hunk 2 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.Console/OutputConsole/OutputConsole.cs:90]: fixed in next commit.

File: src/NuGet.Clients/NuGet.Console/OutputConsole/SharedOutputConsole.cs (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.Console/OutputConsole/SharedOutputConsole.cs]: fixed in next commit.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 4 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: Nice catch. Yeah it should be awaited.

Comment (nkolev92) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: Fixed.

**Pull request merged**